File: src/main/java/no/nav/syfo/consumer/ws/PersonConsumer.java
package no.nav.syfo.consumer.ws;

import lombok.extern.slf4j.Slf4j;
import no.nav.tjeneste.virksomhet.person.v3.binding.HentGeografiskTilknytningPersonIkkeFunnet;
import no.nav.tjeneste.virksomhet.person.v3.binding.HentGeografiskTilknytningSikkerhetsbegrensing;
import no.nav.tjeneste.virksomhet.person.v3.binding.PersonV3;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.*;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentGeografiskTilknytningRequest;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentGeografiskTilknytningResponse;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentPersonnavnBolkRequest;
import org.springframework.stereotype.Component;

import javax.inject.Inject;

import static java.util.Optional.of;
import static java.util.Optional.ofNullable;
import static net.logstash.logback.encoder.org.apache.commons.lang.WordUtils.capitalizeFully;
import static org.springframework.util.StringUtils.isEmpty;

@Component
@Slf4j
public class PersonConsumer {

    private PersonV3 personV3;

    @Inject
    public PersonConsumer(PersonV3 personV3) {
        this.personV3 = personV3;
    }


    public String finnBrukerPersonnavnByFnr(String fnr) {
        return personV3.hentPersonnavnBolk(new HentPersonnavnBolkRequest()
                .withAktoerListe(new PersonIdent().withIdent(new NorskIdent().withIdent(fnr))))
                .getAktoerHarNavnListe()
                .stream()
                .map(AktoerHarNavn::getPersonnavn)
                .map(this::fulltNavn)
                .findFirst()
                .orElse(null);
    }

    public GeografiskTilknytning hentGeografiskTilknytning(String fnr) {
        try {
            HentGeografiskTilknytningResponse response = personV3.hentGeografiskTilknytning(new HentGeografiskTilknytningRequest()
                    .withAktoer(new PersonIdent().withIdent(new NorskIdent().withIdent(fnr))));

            return GeografiskTilknytning
                    .builder()
                    .geografiskTilknytning(of(response.getGeografiskTilknytning()).map(geografiskTilknytning -> geografiskTilknytning.getGeografiskTilknytning()).orElse(null))
                    .diskresjonskode(ofNullable(response.getDiskresjonskode()).map(Kodeverdi::getValue).orElse(null))
                    .build();
        } catch (HentGeografiskTilknytningSikkerhetsbegrensing | HentGeografiskTilknytningPersonIkkeFunnet e) {
            log.error("Feil ved henting av geografisk tilknytning", e);
            throw new RuntimeException("Feil ved henting av geografisk tilknytning", e);
        }
    }

    private String fulltNavn(Personnavn personnavn) {
        String navn;
        if (isEmpty(personnavn.getFornavn())) {
            navn = personnavn.getEtternavn();
        } else if (isEmpty(personnavn.getMellomnavn())) {
            navn = (personnavn.getFornavn() + " " + personnavn.getEtternavn());
        } else {
            navn = (personnavn.getFornavn() + " " + personnavn.getMellomnavn() + " " + personnavn.getEtternavn());
        }

        char[] delimiters = {' ', '-'};
        return capitalizeFully(navn, delimiters);
    }
}

File: src/main/kotlin/no/nav/syfo/consumer/ws/PersonConsumer.kt
package no.nav.syfo.consumer.ws

import net.logstash.logback.encoder.org.apache.commons.lang.WordUtils.capitalizeFully
import no.nav.syfo.log
import no.nav.tjeneste.virksomhet.person.v3.binding.HentGeografiskTilknytningPersonIkkeFunnet
import no.nav.tjeneste.virksomhet.person.v3.binding.HentGeografiskTilknytningSikkerhetsbegrensing
import no.nav.tjeneste.virksomhet.person.v3.binding.PersonV3
import no.nav.tjeneste.virksomhet.person.v3.informasjon.NorskIdent
import no.nav.tjeneste.virksomhet.person.v3.informasjon.PersonIdent
import no.nav.tjeneste.virksomhet.person.v3.informasjon.Personnavn
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentGeografiskTilknytningRequest
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentPersonnavnBolkRequest
import org.springframework.stereotype.Component
import javax.inject.Inject

@Component
class PersonConsumer @Inject
constructor(private val personV3: PersonV3) {


    fun finnBrukerPersonnavnByFnr(fnr: String): String {
        return personV3.hentPersonnavnBolk(HentPersonnavnBolkRequest()
                .withAktoerListe(PersonIdent().withIdent(NorskIdent().withIdent(fnr))))
                .aktoerHarNavnListe
                .map { it.personnavn }
                .map { fulltNavn(it) }
                .first()
    }

    fun hentGeografiskTilknytning(fnr: String): GeografiskTilknytning {
        try {
            val response = personV3.hentGeografiskTilknytning(HentGeografiskTilknytningRequest()
                    .withAktoer(PersonIdent()
                    .withIdent(NorskIdent()
                    .withIdent(fnr))))

            return GeografiskTilknytning(response.geografiskTilknytning?.geografiskTilknytning, response.diskresjonskode?.value)


        } catch (e: HentGeografiskTilknytningSikkerhetsbegrensing) {
            log().error("Feil ved henting av geografisk tilknytning", e)
            throw RuntimeException("Feil ved henting av geografisk tilknytning", e)
        } catch (e: HentGeografiskTilknytningPersonIkkeFunnet) {
            log().error("Feil ved henting av geografisk tilknytning", e)
            throw RuntimeException("Feil ved henting av geografisk tilknytning", e)
        }

    }

    private fun fulltNavn(personnavn: Personnavn): String {
        val navn: String = when {
            personnavn.fornavn.isNullOrBlank() -> personnavn.etternavn
            personnavn.mellomnavn.isNullOrBlank() -> personnavn.fornavn + " " + personnavn.etternavn
            else -> personnavn.fornavn + " " + personnavn.mellomnavn + " " + personnavn.etternavn
        }

        val delimiters = charArrayOf(' ', '-')
        return capitalizeFully(navn, delimiters)
    }
}
